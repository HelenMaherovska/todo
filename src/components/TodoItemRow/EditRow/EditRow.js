import React from "react";
import PropTypes from "prop-types";

const EditRow = props => {
  const { text, id, onToggleEdit, onEditTodo, setRef } = props;
  return (
    <tr>
      <td>
        <input
          defaultValue={text}
          type="text"
          ref={setRef}
          className="form-control todo-text"
        />
      </td>
      <td />
      <td>
        <button
          onClick={() => onEditTodo(id)}
          type="button"
          className="btn btn-success mr-2"
        >
          Save
        </button>
        <button
          onClick={onToggleEdit}
          type="button"
          className="btn btn-danger mr-2"
        >
          Cancel
        </button>
      </td>
    </tr>
  );
};

EditRow.propTypes = {
  id: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
  onEditTodo: PropTypes.func.isRequired,
  onToggleEdit: PropTypes.func.isRequired,
  setRef: PropTypes.func.isRequired
};

export default EditRow;
