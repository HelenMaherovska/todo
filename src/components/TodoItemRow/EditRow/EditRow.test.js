import React from "react";
import { shallow } from "enzyme";
import EditRow from "components/TodoItemRow/EditRow/EditRow";

describe("editRow", () => {
  const id = "12345";
  const text = "Learn unit testing";
  const mockToggleEdit = jest.fn();
  const mockEditTodo = jest.fn();
  const mockSetRef = jest.fn();

  const props = {
    id,
    text,
    onToggleEdit: mockToggleEdit,
    onEditTodo: mockEditTodo,
    setRef: mockSetRef
  };
  const editRow = shallow(<EditRow {...props} />);

  it("renders correctly", () => {
    expect(editRow).toMatchSnapshot();
  });

  describe("edit markup", () => {
    it("has the input control", () => {
      expect(editRow.find(".form-control")).toHaveLength(1);
    });
    it("has the input control", () => {
      expect(editRow.find(".form-control")).toHaveLength(1);
    });

    it("has the `Save` button", () => {
      expect(editRow.find(".btn-success").text()).toEqual("Save");
    });

    it("has the `Cancel` button", () => {
      expect(editRow.find(".btn-danger").text()).toEqual("Cancel");
    });

    describe("when clicking the `Save` button", () => {
      beforeEach(() => {
        editRow.find(".btn-success").simulate("click");
      });

      it("fires the onEditTodo function and saves the changes", () => {
        expect(mockEditTodo).toHaveBeenCalledWith(id);
      });
    });

    describe("when clicking the `Cancel` button", () => {
      beforeEach(() => {
        editRow.find(".btn-danger").simulate("click");
      });

      it("fires the onToggleEdit function and changes the markup", () => {
        expect(mockToggleEdit).toHaveBeenCalled();
      });
    });
  });
});
