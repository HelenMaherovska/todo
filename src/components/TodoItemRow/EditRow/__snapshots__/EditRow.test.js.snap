// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`editRow renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EditRow
    id="12345"
    onEditTodo={[MockFunction]}
    onToggleEdit={[MockFunction]}
    setRef={[MockFunction]}
    text="Learn unit testing"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <td>
          <input
            className="form-control todo-text"
            defaultValue="Learn unit testing"
            type="text"
          />
        </td>,
        <td />,
        <td>
          <button
            className="btn btn-success mr-2"
            onClick={[Function]}
            type="button"
          >
            Save
          </button>
          <button
            className="btn btn-danger mr-2"
            onClick={[MockFunction]}
            type="button"
          >
            Cancel
          </button>
        </td>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <input
            className="form-control todo-text"
            defaultValue="Learn unit testing"
            type="text"
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "form-control todo-text",
            "defaultValue": "Learn unit testing",
            "type": "text",
          },
          "ref": [MockFunction],
          "rendered": null,
          "type": "input",
        },
        "type": "td",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "td",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              className="btn btn-success mr-2"
              onClick={[Function]}
              type="button"
            >
              Save
            </button>,
            <button
              className="btn btn-danger mr-2"
              onClick={[MockFunction]}
              type="button"
            >
              Cancel
            </button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Save",
              "className": "btn btn-success mr-2",
              "onClick": [Function],
              "type": "button",
            },
            "ref": null,
            "rendered": "Save",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Cancel",
              "className": "btn btn-danger mr-2",
              "onClick": [MockFunction],
              "type": "button",
            },
            "ref": null,
            "rendered": "Cancel",
            "type": "button",
          },
        ],
        "type": "td",
      },
    ],
    "type": "tr",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <td>
            <input
              className="form-control todo-text"
              defaultValue="Learn unit testing"
              type="text"
            />
          </td>,
          <td />,
          <td>
            <button
              className="btn btn-success mr-2"
              onClick={[Function]}
              type="button"
            >
              Save
            </button>
            <button
              className="btn btn-danger mr-2"
              onClick={[MockFunction]}
              type="button"
            >
              Cancel
            </button>
          </td>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <input
              className="form-control todo-text"
              defaultValue="Learn unit testing"
              type="text"
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "form-control todo-text",
              "defaultValue": "Learn unit testing",
              "type": "text",
            },
            "ref": [MockFunction],
            "rendered": null,
            "type": "input",
          },
          "type": "td",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "td",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                className="btn btn-success mr-2"
                onClick={[Function]}
                type="button"
              >
                Save
              </button>,
              <button
                className="btn btn-danger mr-2"
                onClick={[MockFunction]}
                type="button"
              >
                Cancel
              </button>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Save",
                "className": "btn btn-success mr-2",
                "onClick": [Function],
                "type": "button",
              },
              "ref": null,
              "rendered": "Save",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Cancel",
                "className": "btn btn-danger mr-2",
                "onClick": [MockFunction],
                "type": "button",
              },
              "ref": null,
              "rendered": "Cancel",
              "type": "button",
            },
          ],
          "type": "td",
        },
      ],
      "type": "tr",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
